heat-eq-simulation/1D_CN.py

```diff
@@ -0,0 +1,88 @@
+# Crank-Nicolson method is an implicit numerical method
+# for derving the solution to PDEs
+
+# we will use the crank-nicolson method below to simulate the 1D Heat Equation
+import numpy
+import matplotlib.pyplot as pyPlot
+from scipy.sparse import diags
+from scipy.sparse.linalg import spsolve
+
+# Global Constants
+
+lengthOfRod = 10
+maxTime = 1
+diffusivityConstant = 1
+numPointsSpace = 1000
+numPointsTime = 2000
+
+# Length Vector plotted on x-axis
+xDomain = numpy.linspace(0, lengthOfRod, numPointsSpace)
+# Time Vector plotted on y-axis
+timeDomain = numpy.linspace(0, maxTime, numPointsTime)
+
+timeStepSize = timeDomain[1] - timeDomain[0]
+spaceStepSize = xDomain[1] - xDomain[0]
+
+# lambda functions for u0- initial condition and alpha and beta- boundary condtions
+u0 = lambda x: numpy.sin(x)
+alpha = lambda t: 5 * t
+beta = lambda t: numpy.sin(lengthOfRod) + 2*t
+
+#intialiizing the boundary conditions and initial condtions
+boundaryConditions = numpy.array([alpha(timeDomain), beta(timeDomain)])
+intialConditions = u0(xDomain)
+
+# error assertion for intial nd boundary conditions
+eps = 1e-12
+err = numpy.abs(u0(0) - alpha(0))
+assert(err < eps)
+
+# Empty Matrix/NestedList with zeroes
+tempMatrix = numpy.zeros((numPointsSpace, numPointsTime))
+tempMatrix[0,:] = boundaryConditions[0]
+tempMatrix[-1,:] = boundaryConditions[1]
+tempMatrix[:, 0] = intialConditions
+
+lambdaConstant = (diffusivityConstant * timeStepSize) / (2*spaceStepSize**2)
+print(lambdaConstant)
+
+# Set up tridiagonal matrix coefficients
+mainDiagonal = (1 + 2 * lambdaConstant) * numpy.ones(numPointsSpace - 2)
+lowerDiagonal = -lambdaConstant * numpy.ones(numPointsSpace - 3)
+upperDiagonal = -lambdaConstant * numpy.ones(numPointsSpace - 3)
+
+# Create the sparse tridiagonal matrix A
+A = diags([lowerDiagonal, mainDiagonal, upperDiagonal], offsets=[-1, 0, 1], format='csr')
+#pyPlot.spy(A)
+#pyPlot.show()
+
+# Time-stepping loop
+for tau in range(1, numPointsTime):
```

Comment:
At a glance this looks good, but have you tested to see whether it produces (approximately) the same solution as the explicit method (for the same initial/boundary conditions)?

Reply:
@alphaville To the naked eye- yes, the 3d plot looks almost the same. Is there any specific method you have in mind that I should employ to verify this? The initial and boundary condition are same on both my scripts, since, I just copied them over from the ftcs explicit script

Reply:
You can just compare the solutions. Use the exact same initial conditions and grid step sizes. Each method gives you a matrix $A_{\tau, j}$; let's call these $A^{\rm xpl}$ (for the explicit method) and $A^{\rm cn}$ (for the Crank-Nicolson method). You can compute
$$e = \max_{\tau, j}|A_{\tau, j}^{\rm xpl} - A_{\tau, j}^{\rm nc}|,$$
and see if $e$ is reasonably small. You can make a unit test to check whether for example $e< 10^{-3}$. This should hold provided you have chosed adequately small step sizes ($\delta x$ and $\delta t$).

As you said, we can merge this now and then you can create a new PR.